packages/flutter/lib/src/painting/image_provider.dart

```diff
@@ -322,17 +331,36 @@ abstract class ImageProvider<T> {
       }
       key.then<void>((T key) {
         obtainedKey = key;
-        final ImageStreamCompleter completer = PaintingBinding.instance.imageCache.putIfAbsent(
-          key,
-          () => load(key, PaintingBinding.instance.instantiateImageCodec),
-          onError: handleError,
-        );
-        if (completer != null) {
-          stream.setCompleter(completer);
+        try {
+          useKey(configuration, stream, key, handleError);
+        } catch (error, stackTrace) {
+          handleError(error, stackTrace);
         }
       }).catchError(handleError);
     });
-    return stream;
+  }
+
+  /// Called when [resolve] has finished calling [obtainKey].
```

Comment:
From the docs and method name its not clear what useKey is supposed to do if I were to override it.